diff --git a/node_modules/hardhat/dist/src/internal/builtin-plugins/solidity/build-system/compilation-job.js b/node_modules/hardhat/dist/src/internal/builtin-plugins/solidity/build-system/compilation-job.js
index c250a14..74724b8 100644
--- a/node_modules/hardhat/dist/src/internal/builtin-plugins/solidity/build-system/compilation-job.js
+++ b/node_modules/hardhat/dist/src/internal/builtin-plugins/solidity/build-system/compilation-job.js
@@ -75,8 +75,28 @@ export class CompilationJobImplementation {
                 "": ["ast"],
             },
         };
-        // TODO: Deep merge the user output selection with the default one
+        
         const outputSelection = defaultOutputSelection;
+        const configOutputSelection = settings.outputSelection ?? {};
+    
+        for (const fileKey of Object.keys(configOutputSelection).sort()) {
+          outputSelection[fileKey] ??= {};
+          const configFileOutputSelection = configOutputSelection[fileKey] ?? {};
+    
+          for (const contractKey of Object.keys(configFileOutputSelection).sort()) {
+            outputSelection[fileKey][contractKey] ??= [];
+            const configContractOutputSelection =
+              configFileOutputSelection[contractKey] ?? [];
+    
+            outputSelection[fileKey][contractKey] = Array.from(
+              new Set([
+                ...outputSelection[fileKey][contractKey],
+                ...configContractOutputSelection,
+              ]),
+            ).sort();
+          }
+        }
+    
         return {
             language: "Solidity",
             settings: {
